Add support of count_include_pad and test end to end test for AveragePool

dwarakrajagopal opened the pull request:
Add support of count_include_pad end to end test for AveragePool

We can export AveragePool from PyTorch with count_include_pad attribute. However, we don't directly support it in Caffe2's ONNX backend. 
We also want to check whether we can pass the end to end test for average pool operator with count_include_pad attribute (pytorch => onnx => caffe2)

Commit:
Support for count_include_pad

dwarakrajagopal committed:
Add support of count_include_pad test end to end for AveragePool

dwarakrajagopal committed:
cleanup dead code

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -664,11 +664,14 @@ def test_maxpool2d_single_padding(self):
 
-    @unittest.skip("C2 and PyTorch have small difference in padding implementation")
     def test_avgpool2d(self):
         model = nn.AvgPool2d(5, padding=(2))
```
I think the default value of count_include_pad is True, we can set it as True in this case.

dwarakrajagopal replied:
Done.

houseroad commented on caffe2/onnx/backend.cc:
```diff
@@ -548,50 +548,12 @@ Caffe2Ops Caffe2Backend::CreateConvPoolOpBase(
-  auto& attributes = onnx_node->attributes;
   Caffe2Ops ret;
   // Pad
   bool padding = false;
```
Let's also remove this logic.

dwarakrajagopal replied:
Removing the entire CreatePadPool function and use CreateConvPoolOpBase directly for AveragePool and GlobalAveragePool

dwarakrajagopal committed:
Add support of count_include_pad and test end to end test for AveragePool

Summary:
Add support of count_include_pad end to end test for AveragePool
We can export AveragePool from PyTorch with count_include_pad attribute. However, we don't directly support it in Caffe2's ONNX backend.
We also want to check whether we can pass the end to end test for average pool operator with count_include_pad attribute (pytorch => onnx => caffe2).

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -664,9 +664,12 @@ def test_maxpool2d_single_padding(self):
-    @unittest.skip("C2 and PyTorch have small difference in padding implementation")
     def test_avgpool2d(self):
-        model = nn.AvgPool2d(5, padding=(2))
+        model = nn.AvgPool2d(5, padding=(2), count_include_pad=True)
```
I think this should be false.

dwarakrajagopal replied:
Re-writing this via 3 separate test cases for clarity - 1)avg_pool2d_with_count_include_pad_set_true; 2)avg_pool2d_with_count_include_pad_set_false; 3)avg_pool2d

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -664,9 +664,12 @@ def test_maxpool2d_single_padding(self):
+        self.run_model_test(model, train=False, batch_size=BATCH_SIZE)
+
+    def test_avgpool2d_count_include_pad(self):
+        model = nn.AvgPool2d(5, padding=(2), count_include_pad=False)
```
And this one should be true, :-)

dwarakrajagopal replied:
Re-writing this via 3 separate test cases for clarity - 1)avg_pool2d_with_count_include_pad_set_true; 2)avg_pool2d_with_count_include_pad_set_false; 3)avg_pool2d

dwarakrajagopal committed:
add 3 specific tests for different versions of AvgPool - default, with and without count_include_pad

houseroad reviewed (approved):
Looks good.
